src/split.cc

```diff
@@ -0,0 +1,155 @@
+	}
+	return {small_chunk_size, large_chunk_size, num_large_chunks};
+}
+
+} // namespace
+
+namespace celerity::detail {
+
+std::vector<chunk<3>> split_1d(const chunk<3>& full_chunk, const range<3>& granularity, const size_t num_chunks) {
+#ifndef NDEBUG
+	assert(num_chunks > 0);
+	for(int d = 0; d < 3; ++d) {
+		assert(granularity[d] > 0);
+		assert(full_chunk.range[d] % granularity[d] == 0);
+	}
+#endif
+
+	// Due to split granularity requirements or if num_workers > global_size[0],
+	// we may not be able to create the requested number of chunks.
+	const std::array<size_t, 1> actual_num_chunks = {std::min(num_chunks, full_chunk.range[0] / granularity[0])};
+	const auto [small_chunk_size, large_chunk_size, num_large_chunks] = compute_small_and_large_chunks<1>(full_chunk, granularity, actual_num_chunks);
+
+	std::vector<chunk<3>> result(actual_num_chunks[0], {full_chunk.offset, full_chunk.range, full_chunk.global_size});
+	for(auto i = 0u; i < num_large_chunks[0]; ++i) {
+		result[i].range[0] = large_chunk_size[0];
+		result[i].offset[0] += i * large_chunk_size[0];
+	}
+	for(auto i = num_large_chunks[0]; i < actual_num_chunks[0]; ++i) {
+		result[i].range[0] = small_chunk_size[0];
+		result[i].offset[0] += num_large_chunks[0] * large_chunk_size[0] + (i - num_large_chunks[0]) * small_chunk_size[0];
+	}
+
+#ifndef NDEBUG
+	sanity_check_split(full_chunk, result);
+#endif
+
+	return result;
+}
+
+// TODO: Make the split dimensions configurable for 3D chunks?
+std::vector<chunk<3>> split_2d(const chunk<3>& full_chunk, const range<3>& granularity, const size_t num_chunks) {
+#ifndef NDEBUG
+	assert(num_chunks > 0);
+	for(int d = 0; d < 3; ++d) {
+		assert(granularity[d] > 0);
+		assert(full_chunk.range[d] % granularity[d] == 0);
+	}
+#endif
+
+	const auto assign_factors = [&full_chunk, &granularity, &num_chunks](const size_t factor) {
+		assert(num_chunks % factor == 0);
+		const size_t max_chunks[2] = {full_chunk.range[0] / granularity[0], full_chunk.range[1] / granularity[1]};
+		const size_t f0 = factor;
+		const size_t f1 = num_chunks / factor;
+
+		// Decide in which direction to split by first checking which
+		// factor assignment produces more chunks under the given constraints.
+		const std::array<size_t, 2> split0 = {std::min(f0, max_chunks[0]), std::min(f1, max_chunks[1])};
+		const std::array<size_t, 2> split1 = {std::min(f1, max_chunks[0]), std::min(f0, max_chunks[1])};
+		const auto count0 = split0[0] * split0[1];
+		const auto count1 = split1[0] * split1[1];
+
+		if(count0 == count1) {
+			// If we're tied for the number of chunks we can create, try some heuristics to decide.
+
+			// If domain is square(-ish), prefer splitting along slower dimension.
+			// (These bounds have been chosen arbitrarily!)
+			const double squareishness = std::sqrt(full_chunk.range.size()) / static_cast<double>(full_chunk.range[0]);
+			if(squareishness > 0.95 && squareishness < 1.05) { return (f0 >= f1) ? split0 : split1; }
```

Comment:
`f0` and `f1` are the two split factors, i.e., how many cuts there are on each side of the domain. E.g. for 16 nodes `f0 = f1 = 4`, and the choice doesn't matter. But for 8 nodes `f0 = 2` and `f1 = 4`, so there's a choice to make. In this case, if the domain is about square, it will assign `f1` (more cuts) to dimension 0 (the slower dimension), and `f0` (fewer cuts) to dimension 1.

Reply:
I understand what's happening, but some of that explanation could be code comments. Also consider making some of the variable names more _wortreich_ to convey that `0` and `1` refer to the split direction each time.

Reply:
It should be more understandable now hopefully!